### PR TITLE
Bulk import CSV choices into a rating allocation

### PR DESCRIPTION
**Pull request by DrCuriosity**
**Note**: this PR builds upon the group restriction feature offered in #224; please review that first.

This allows the user to upload a CSV file containing data for multiple choices to a rating allocation activity.

When editing choices, an "Upload CSV" button is displayed while rating has not yet started.

image

This takes the administrator to an upload form:

image

They can choose test their file using the "Test upload" checkbox, which will do a dry run of the import without committing values to the database. Otherwise, new choices and group restriction (if any) will be added to the activity.

A new `upload_choices` action is added in locallib.php to process the new `form_upload_choices.php` form. The CSV import process has been encapsulated in `classes/choice_importer.php`, employing existing functions from the `ratingallocate` class as appropriate.

The user is issued warning notifications about data errors during the import (or import test) up to a `MAX_WARNING_COUNT` constant.

Testing the CSV importer is done in `tests/mod_ratingallocate_choice_importer_test.php`

### Related issues:
* #224 (dependency)
* #16 (historical)

### CODE REVIEW

**Comment by DrCuriosity** [comment]
Branch has been rebased to adjust for recommended changes on prior feature, per [#224 (comment)](https://github.com/learnweb/moodle-mod_ratingallocate/pull/224#issuecomment-909512761).

File: classes/choice_importer.php
Comment (PhMemmel) [inline classes/choice_importer.php:185]: Need to use a different delimiter here, e. g. ";", as comma is already used as csv delimiter

Comment (danmarsden) [reply classes/choice_importer.php:185]: thanks for the review - as mentioned in #224 I'd prefer to wait for the plugin maintainers to provide feedback on any changes they would like to see before we come back to this again.

**Comment by danmarsden** [comment]
I'm happy to look at resolving conflicts here if we could get some feedback on this PR.

**Comment by PhMemmel** [comment]
Let me just provide some additional feedback here:

We've been running this PR and the group restriction feature together with some further improvements (public fork: https://github.com/mebis-lp/moodle-mod_ratingallocate) for quite some time now and can say that it works very well. Thank you for providing these PRs @danmarsden!

If these land, we of course will be happy to also provide PRs with our own improvements (or just feel free to cherry pick from our fork, of course).

File: classes/choice_importer.php
Comment (Laur0r) [reply classes/choice_importer.php:185]: I agree, otherwise it is not possible to specify more than one group as the comma is interpreted as another column for the csv import.

File: classes/choice_importer.php
Comment (Laur0r) [inline classes/choice_importer.php:180]: This should be trimmed for whitespaces first, otherwise a trailing whitespace will lead to a choice with usegroups enabled without an assigned group.

**Review by Laur0r — CHANGES_REQUESTED** [review]
Thank you very much for your contribution and sorry for the late reply! When the changes are made to #242, this should be rebased.

File: classes/choice_importer.php
Comment (Laur0r) [reply classes/choice_importer.php:185]: I didn't know about the quotation marks, thanks for pointing that out! This underlines my point that the desired formatting for the CSV File should be communicated clearly  to the user e.g by examples. However, it may still be more intuitive to use a different separator here so that the quotation marks are not necessary for multiple groups.

**Review by PhMemmel — COMMENTED** [review]
Hi @Laur0r,

I adressed all the issues you mentioned in #246. I also added a commit on top, because unit tests are failing under current master due to a deprecation warning. Let me know if you want to get this fixed in a separate issue/PR.

Another point: Currently the upload choices CSV button is only being shown if rating phase has not started yet. I can understand the idea behind it. However, it is inconsistent with the user being able to add new choices by clicking "Add new choice". So users might wonder why they can still add single choices, but the "Upload CSV" button has gone for some reason.
I would vote to remove the restriction to the "before rating" phase. Sometimes users want to add choices even if rating phase has already been started (for example they manually hid the activity in the course page or added a date restriction on the activity, just use the activity for testing purposes etc.).
If you do not want to remove the restriction I would suggest to apply the restriction to the "Add choice" button to make behavior consistent.

BTW: I think there should be an update to the moodle-ci.yml for GHA. Currently there is only PHP 7.4 running which is already out of support and moodle is supporting at least PHP 8.0 from Moodle 4.0 on. Also Moodle 4.1 is missing and it's maybe even worth to add master branch to spot any errors early.

**Comment by Laur0r** [comment]
Dear @DrCuriosity and @danmarsden, this PR is merged with #246. Thank you for your contribution!

**Comment by danmarsden** [comment]
Great thanks! Nice to see this land!!

**Pull request closed without merging**
